sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -819,6 +819,8 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
       case r: logical.RepartitionByExpression =>
         val shuffleOrigin = if (r.partitionExpressions.isEmpty && r.optNumPartitions.isEmpty) {
           REBALANCE_PARTITIONS_BY_NONE
+        } else if (!r.userSpecified) {
+          REBALANCE_PARTITIONS_BY_COL
```

Comment:
`REBALANCE_PARTITIONS_BY_COL` seems more suitable to be from `RebalancePartitions`? Although there is `REBALANCE_PARTITIONS_BY_NONE` here too.

After rebalancing, do we still guarantee partitioning requirement by the partition expressions? Seems not to me as `RebalancePartitions` said:

> ... It also try its best to partition the child output by `partitionExpressions`....